### ISSUE TITLE
Allow unixd with anonymous disabled

### ISSUE THREAD

**Issue by johansmitsnl** [kanidm/kanidm#3319]
### Is your feature request related to a problem? Please describe

Have the unixd integration with the anonymous account disabled. 

### Describe the solution you'd like

Include a service token or something else to make a authorized connection. Like this you can also limit the devices that have access or if they are compromised to prevent users from logging in.

### Describe alternatives you've considered

No

### Additional context

By default with anonymous you can get a lot of data (ssh, groups, login shell, gidnumber) that can be used to gain specific knowledge of the users environment. And use this information to build phishing attacks. I don't want this information to be in the open.

**Comment by jinnatar** [comment]
First, from previous discussions, I'll add two things:
1. The enumeration problem is a bit worse than just users, the anonymous account can by default also enumerate for example enabled oauth2 clients. Ergo, one can (and I've done it as a PoC) find publicly exposed Kanidm instances on Shodan and query exactly what apps are gated by said instance. So not only can one find users to target, you can also likely identify apps they are actually using to enhance the realism of a campaign, or to find attack vectors. Nothing you can't already do, but it's served on a very convenient platter.
2. The alternative mitigation in addition to disabling or impairing the anonymous account is to use a WAF or capable reverse proxy to limit what API endpoints are accessible from the public internet.
   - It's best to explicitly allow only known needed endpoints externally.
   -  An example is provided below that allows oauth flows & user account management to work from the outside. It does not allow most management calls, meaning that also the kanidm cli tool won't work from behind the rules.
   - Allowlist in regex form applied on the path: `^/v1/(auth|self|credential|jwk|reauth|oauth2)`, everything else under `^/v1/` blocked.

### My 2c on a proper solution:
I agree that personally for me empowering unixd with credentials of some sort generated by Kanidm would be a low effort solution from an admin perspective in a small to moderate environment. Most likely I would create, securely store in a vault & distribute at runtime a single token to all trusted server clients, or even perhaps bother with per-host tokens.

Any kind of more automated or less-trusted environment would benefit from a cert dispenser style strategy, say for example for using a secrets vault to mint & dispense them. My knowledge on that front goes only so far, the canonical example that comes to mind would be using PKI, but that would require Kani to accept certs for auth. The benefit being that PKI is a well tooled solution for large/complex/over-engineered infra, even if quite obtuse. Just mentioning it as an example of how trust distribution has been solved elsewhere, happy to hear of any easier alternatives!

**Comment by Firstyear** [comment]
> First, from previous discussions, I'll add two things:
> 
>     1. The enumeration problem is a bit worse than just users, the anonymous account can by default also enumerate for example enabled oauth2 clients. Ergo, one can (and I've done it as a PoC) find publicly exposed Kanidm instances on Shodan and query exactly what apps are gated by said instance.

Only if you add idm_all_accounts as a valid user of an oauth2 client. We have a special access control so that users can see oauth2 clients they are *allowed* to access. So if you add idm_all_accounts, you of course are saying "anonymous can see these". Might be still a surprise to most people though, so I think we can easily remove this as a special case in the access control code. 

> 
>     2. The alternative mitigation in addition to disabling or impairing the anonymous account is to use a WAF or capable reverse proxy to limit what API endpoints are accessible from the public internet.

I want to limit and remove anonymous in future as we go, so I think we should focus on cleanup. 

> ### My 2c on a proper solution:
> 
> I agree that personally for me empowering unixd with credentials of some sort generated by Kanidm would be a low effort solution from an admin perspective in a small to moderate environment. Most likely I would create, securely store in a vault & distribute at runtime a single token to all trusted server clients, or even perhaps bother with per-host tokens.
> 

Agreed. I think it's probably easier to implement this in the short term. 

> Any kind of more automated or less-trusted environment would benefit from a cert dispenser style strategy, say for example for using a secrets vault to mint & dispense them. My knowledge on that front goes only so far, the canonical example that comes to mind would be using PKI, but that would require Kani to accept certs for auth. The benefit being that PKI is a well tooled solution for large/complex/over-engineered infra, even if quite obtuse. Just mentioning it as an example of how trust distribution has been solved elsewhere, happy to hear of any easier alternatives!

And yes, that's domain join :) which we want to do but is a longer project.

**Comment by yaleman** [comment]
I'd say removing `anonymous` from `idm_all_accounts` would be sensible 😄

**Comment by Firstyear** [comment]
anonymous is *technically* an account though.

**Comment by yaleman** [comment]
Yeah, but it's also technically a magical jerk in a load of ways... 😄

**Comment by Firstyear** [comment]
Woooo edge cases! 

And thus why I want to work to eventually remove it :)

**Comment by pdf** [comment]
So AIUI the only viable choice currently for working unixd is to leave `anonymous` enabled (including anon LDAP binds, etc), allowing user enumeration by anyone who can get a socket to kanidm?

I guess SSSD or similar would be the only alternative currently?

**Comment by Firstyear** [comment]
The next release of Kanidm (1.8) will support a service-account token in unixd to authenticate this service allowing anonymous to be disabled.

**Comment by Blacks-Army** [comment]
> The next release of Kanidm (1.8) will support a service-account token in unixd to authenticate this service allowing anonymous to be disabled.

Hey, is it possible yet? I could not find anything related to that in the docs.

**Comment by Firstyear** [comment]
@Blacks-Army It is possible, just not documented. I will correct this in the coming days :)

**Comment by Blacks-Army** [comment]
> @Blacks-Army It is possible, just not documented. I will correct this in the coming days :) 

Found the PR :) Thanks for the Hint, will try it later today.

**Comment by Firstyear** [comment]
@Blacks-Army short notes:

```
kanidm service-account create unix-bind "Unix Bind" idm_unix_admins

kanidm group add-members idm_unix_authentication_read unix-bind

kanidm service-account api-token generate unix-bind ktest
```

Then in kanidm/unixd

```
service_account_token_path = "/usr/local/etc/kanidm/token"
```

Where `token` has the api-token on a single line. Should be set to root:kanidmd or similar permissions to protect. You can alternately  use systemd with `Environment=KANIDM_SERVICE_ACCOUNT_TOKEN_PATH=/usr/local/etc/kanidm/token` or systemd credentials in a similar way.
